File: lib/swimDuplicates.js
/**
 * Detect duplicate swim sessions (same workout uploaded twice).
 */

const PACE_TOLERANCE_SEC = 5;

const coreMetricsMatch = (a, b) => {
  if (a.date !== b.date) return false;

  const ma = a.metrics || {};
  const mb = b.metrics || {};

  if (ma.distanceM == null || mb.distanceM == null) return false;
  if (ma.durationSec == null || mb.durationSec == null) return false;

  if (ma.distanceM !== mb.distanceM || ma.durationSec !== mb.durationSec) return false;

  if (ma.paceSecPer100m != null && mb.paceSecPer100m != null) {
    if (Math.abs(ma.paceSecPer100m - mb.paceSecPer100m) > PACE_TOLERANCE_SEC) return false;
  }

  if (ma.timeRange && mb.timeRange && ma.timeRange !== mb.timeRange) return false;

  return true;
};

export const findDuplicateSession = (sessions, candidate, excludeId = null) => {
  if (!candidate?.date || !sessions?.length) return null;

  const match = sessions.find(
    (s) => s.id !== excludeId && coreMetricsMatch(candidate, s)
  );

  return match ?? null;
};

export { coreMetricsMatch };

/** @internal */
export const __testing = { coreMetricsMatch, PACE_TOLERANCE_SEC };

File: ios/AapSC/Lib/SwimDuplicates.swift
import Foundation

enum SwimDuplicates {
    static let paceToleranceSec = 5

    static func coreMetricsMatch(_ a: SwimSession, _ b: SwimSession) -> Bool {
        guard a.date == b.date else { return false }
        return coreMetricsMatch(date: a.date, metricsA: a.metrics, metricsB: b.metrics)
    }

    static func findDuplicateSession(
        _ sessions: [SwimSession],
        candidate: SwimSession,
        excludeId: String? = nil
    ) -> SwimSession? {
        guard !candidate.date.isEmpty, !sessions.isEmpty else { return nil }
        if let workoutUUID = candidate.healthKitWorkoutUUID,
           let match = sessions.first(where: { session in
               session.id != excludeId && session.healthKitWorkoutUUID == workoutUUID
           }) {
            return match
        }
        return sessions.first { session in
            session.id != excludeId && coreMetricsMatch(candidate, session)
        }
    }

    private static func coreMetricsMatch(date: String, metricsA: SwimMetrics, metricsB: SwimMetrics) -> Bool {
        guard metricsA.distanceM != nil, metricsB.distanceM != nil else { return false }
        guard metricsA.durationSec != nil, metricsB.durationSec != nil else { return false }
        guard metricsA.distanceM == metricsB.distanceM else { return false }
        guard metricsA.durationSec == metricsB.durationSec else { return false }

        if let paceA = metricsA.paceSecPer100m, let paceB = metricsB.paceSecPer100m {
            if abs(paceA - paceB) > paceToleranceSec { return false }
        }

        if !metricsA.timeRange.isEmpty, !metricsB.timeRange.isEmpty,
           metricsA.timeRange != metricsB.timeRange {
            return false
        }

        return true
    }
}
